### PR TITLE
quic: fix quic platform test TSAN failure

### PR DESCRIPTION
**Pull request by danzh2010**
Signed-off-by: Dan Zhang <danzh@google.com>

Commit Message: QuicThread test fails under TSAN because of racing between vtable pointer change and calling into its virtual method. Fix it by waiting till the virtual method finishes before destruction.

This is a test setup issue.

Risk Level: low, test only
Testing: N/A
Docs Changes: N/A
Release Notes: N/A
Platform Specific Features: N/A

### CODE REVIEW

File: test/common/quic/platform/quic_platform_test.cc [hunk 1]
```diff
@@ -207,12 +207,24 @@ TEST_F(QuicPlatformTest, QuicThread) {
...
+    void Run() override {
+      std::unique_lock<std::mutex> lk(m_);
+      *value_ += increment_;
+      cv_.notify_one();
+    }
 
   private:
     int* value_;
     int increment_;
+    std::mutex m_;
```
Comment (yanavlasov) [inline test/common/quic/platform/quic_platform_test.cc:226]: The convention is to use absl::Mutex which supports conditional usage as well.

**Review by yanavlasov — CHANGES_REQUESTED** [review]
/wait

File: test/common/quic/platform/quic_platform_test.cc (hunk 1 above)
Comment (danzh2010) [reply test/common/quic/platform/quic_platform_test.cc]: switched to absl mutex now!

**Review by yanavlasov — APPROVED** [review]

**Pull request merged**